Prevent supervisor from deleting itself

roman-mazur opened the pull request:
With recent changes to local mode behavior, supervisor removes all engine objects
that were not present before entering local mode.
With this change, supervisor also detects its own resources and does not delete them
if it was updated during local mode operations.

Related to https://github.com/balena-os/meta-balena/pull/1590 and https://github.com/balena-os/balena-engine/issues/173

Change-type: patch
Signed-off-by: Roman Mazur <roman@balena.io>

roman-mazur commented:
Keeping this as draft until the arch call.

roman-mazur commented:
@balena-ci retest

roman-mazur commented:
Promoting it to "Ready for review".
Exposing container ID without bind mount is tracked in balena-os/balena-engine#173

roman-mazur commented:
Rebased.
@balena-ci retest please

roman-mazur commented:
Got a test failure
https://ci.balena-dev.com/builds/730541
```
1) deviceState
       allows triggering applying the target state:
     Uncaught AssertionError: expected applyTarget to have been called with arguments { force: true, initial: false }
      at Timeout._onTimeout (build/test/05-device-state.spec.js:358:55)
```

roman-mazur commented:
@balena-ci retest once more please

roman-mazur commented:
@CameronDiver we seem to have a flaky test. It succeeded on re-run:
https://ci.balena-dev.com/builds/730719

Have you seen this failure before?

roman-mazur commented:
Rebased again.
@balena-ci retest please

roman-mazur commented:
@balena-ci retest please

roman-mazur commented on src/lib/constants.ts:
```diff
@@ -9,6 +9,7 @@ const constants = {
 	rootMountPoint,
 	databasePath:
 		checkString(process.env.DATABASE_PATH) || '/data/database.sqlite',
+	containerId: checkString(process.env.SUPERVISOR_CONTAINER_ID) || undefined,
```
@CameronDiver @robertgzr this is the env variable we'll use to set the CID with `--env-cid` parameter in `balena run`.

I didn't use BALENA_ prefix, since we don't use it for other variables.

CameronDiver replied:
As long as this is definable in the engine, that's fine by me.

roman-mazur commented:
@balena-ci retest

CameronDiver commented on src/local-mode.ts:
```diff
@@ -142,6 +147,39 @@ export class LocalModeManager {
+			if (this.containerId !== undefined) {
+				// Inspect operation fails (container ID is out of sync?).
+				const fallback = SUPERVISOR_CONTAINER_NAME_FALLBACK;
+				log.error(
```
```suggestion
				log.warn(
```

CameronDiver commented on src/local-mode.ts:
```diff
@@ -142,6 +147,39 @@ export class LocalModeManager {
+				);
+				return this.collectContainerResources(fallback);
+			}
+			throw e;
```
Can you make this a typed error please? Often errors can come from seemingly nowhere, so wrapping this would help know it's origin.

roman-mazur replied:
The problem here is that I don't think dockerode/docker-modem throws a typed error.
https://github.com/apocas/docker-modem/blob/master/lib/modem.js#L257

roman-mazur replied:
Or do you mean wrapping it with our own type?

CameronDiver replied:
Wrapping it with our type, for example
```typescript
import * as TypedError from 'typed-error';
export class SupervisorContainerNotFoundError extends TypedError {}

throw new SupervisorContainerNotFoundError(e)
```

roman-mazur replied:
normal typescript import does not work (lint fails), so I used an approach from other files (like `src/lib/errors.ts`):
```
import TypedError = require('typed-error');
```
I believe the package needs an upgrade, but this should be done in a separate PR.

CameronDiver replied:
Ah yeah I forgot about that - I was waiting for 100% typescript before doing that.

CameronDiver commented on test/local-mode.ts:
```diff
@@ -1,6 +1,7 @@
 import * as assert from 'assert';
```
This fine should be named `local-mode.spec.ts`

CameronDiver commented on test/local-mode.ts:
```diff
@@ -1,6 +1,7 @@
 import * as assert from 'assert';
 import { expect } from 'chai';
 import * as Docker from 'dockerode';
+import * as fs from 'fs';
```
This is unused, and for future reference, we tend to use `import { fs } from 'mz';` to get the free promisified functions.

roman-mazur commented:
@balena-ci retest please

CameronDiver commented on src/local-mode.ts:
```diff
@@ -56,6 +58,12 @@ export class EngineSnapshotRecord {
+const SUPERVISOR_CONTAINER_NAME_FALLBACK = 'resin_supervisor';
+
+/** An error thrown if our own container cannot be inspected. */
+class SupervisorContainerNotFoundError extends TypedError {}
```
Please add this to `./errors.ts`

roman-mazur replied:
Done.

roman-mazur commented:
@balena-ci retest please

CameronDiver commented on src/local-mode.ts:
```diff
@@ -142,6 +148,39 @@ export class LocalModeManager {
+		return new EngineSnapshot(
+			[inspectInfo.Id],
+			[inspectInfo.Image],
+			inspectInfo.Mounts.filter(m => m.Name !== undefined).map(m => m.Name!),
```
I think this should be `== null`. The docker daemon can sometimes return `null` for various fields, or omit the field altogether. The best thing to do with using the docker api is to always use `== null` to cover both instances.

roman-mazur replied:
Good to know, thanks.

roman-mazur replied:
Changed.

CameronDiver commented on src/local-mode.ts:
```diff
@@ -142,6 +148,39 @@ export class LocalModeManager {
+			[inspectInfo.Id],
+			[inspectInfo.Image],
+			inspectInfo.Mounts.filter(m => m.Name !== undefined).map(m => m.Name!),
+			_.values(inspectInfo.NetworkSettings.Networks).map(ns => ns.NetworkID),
```
```suggestion
			_.mapValues(inspectInfo.NetworkSettings.Networks, 'NetworkID'),
```

roman-mazur replied:
It's not type safe! :)
Do you insist?

CameronDiver replied:
Hm I thought it was type safe? Otherwise I'd prefer `mapValues` with a lambda

Page- replied:
I'd go for
```suggestion
			_.mapValues(inspectInfo.NetworkSettings.Networks, ns => ns.NetworkID),
```
to get the best of both

roman-mazur replied:
In fact, `mapValues` transforms object to an object - it does not return an array
https://lodash.com/docs/4.17.14#mapValues

Page- replied:
Right, I was just looking at the type safe bit, you can use straight `_.map` in that case to have an array returned

roman-mazur replied:
Changed.

Page- commented on src/local-mode.ts:
```diff
@@ -70,6 +75,7 @@ export class LocalModeManager {
 		public docker: Docker,
 		public logger: Logger,
 		public db: Database,
+		private containerId: string | undefined = constants.containerId,
```
Just to check, the `| undefined` here is to allow assigning `constants.containerId` right?  Because if it's expected to be able to explicitly pass `undefined` as the value then that will fail due to the default

roman-mazur replied:
It's supposed to be used either as
```
new LocalModeManager(config, docker, logger, db)
```
from the main code where the manager is created - in such case the ID will be taken from constants (using the environment variable)
or as
```
new LocalModeManager(config, docker, logger, db, 'someSpecificValue')
```
which is used in tests to inject a specific container ID

roman-mazur commented:
>Commit message Filter both null and ommited values should be omitted instead

Has already been fixed. Thanks.

@balena-ci retest please

roman-mazur commented:
@balena-ci retest please

roman-mazur commented:
@CameronDiver I agree it's a sensitive word, so I renamed the test and adjusted commit messages.
Branch will be deleted after the merge, so we should be good.

roman-mazur commented:
@balena-ci retest

roman-mazur commented:
Rebased again
@balena-ci retest please
